File: src/sagacraft/data/io_service.py
"""
I/O service for file operations and data persistence
"""

import json
from pathlib import Path
from typing import Dict, Any, Optional
import logging

from core.services import Service


class IOService(Service):
    """
    Handles all file I/O operations

    Provides:
    - Adventure loading/saving
    - Save game management
    - JSON file operations
    - Path management
    """

    def __init__(self, base_dir: str = "."):
        self.logger = logging.getLogger("IOService")
        self.base_dir = Path(base_dir)
        self.adventures_dir = self.base_dir / "adventures"
        self.saves_dir = self.base_dir / "saves"

    def initialize(self, config: Dict[str, Any]):
        """Initialize the I/O service"""
        # Create directories
        self.adventures_dir.mkdir(parents=True, exist_ok=True)
        self.saves_dir.mkdir(parents=True, exist_ok=True)

        # Override paths from config
        if "adventures_dir" in config:
            self.adventures_dir = Path(config["adventures_dir"])
        if "saves_dir" in config:
            self.saves_dir = Path(config["saves_dir"])

        self.logger.info("I/O service initialized")

    def shutdown(self):
        """Cleanup (nothing needed for basic I/O)"""
        pass

    def load_json(self, file_path: Path) -> Optional[Dict[str, Any]]:
        """
        Load JSON file

        Args:
            file_path: Path to JSON file

        Returns:
            Parsed JSON data or None on error
        """
        try:
            with open(file_path, "r") as f:
                return json.load(f)
        except FileNotFoundError:
            self.logger.error(f"File not found: {file_path}")
            return None
        except json.JSONDecodeError as e:
            self.logger.error(f"Invalid JSON in {file_path}: {e}")
            return None
        except Exception as e:
            self.logger.error(f"Error loading {file_path}: {e}")
            return None

    def save_json(self, file_path: Path, data: Dict[str, Any], indent: int = 2) -> bool:
        """
        Save data to JSON file

        Args:
            file_path: Path to save to
            data: Data to save
            indent: JSON indentation

        Returns:
            True on success, False on error
        """
        try:
            file_path.parent.mkdir(parents=True, exist_ok=True)
            with open(file_path, "w") as f:
                json.dump(data, f, indent=indent)
            return True
        except Exception as e:
            self.logger.error(f"Error saving to {file_path}: {e}")
            return False

    def load_adventure(self, adventure_name: str) -> Optional[Dict[str, Any]]:
        """
        Load an adventure by name

        Args:
            adventure_name: Name of adventure (without .json)

        Returns:
            Adventure data or None
        """
        adventure_path = self.adventures_dir / f"{adventure_name}.json"
        return self.load_json(adventure_path)

    def list_adventures(self) -> list:
        """List available adventures"""
        if not self.adventures_dir.exists():
            return []
        return [f.stem for f in self.adventures_dir.glob("*.json")]

    def save_game(self, save_name: str, game_state: Dict[str, Any]) -> bool:
        """
        Save game state

        Args:
            save_name: Name for the save file
            game_state: State data to save

        Returns:
            True on success
        """
        save_path = self.saves_dir / f"{save_name}.json"
        return self.save_json(save_path, game_state)

    def load_game(self, save_name: str) -> Optional[Dict[str, Any]]:
        """
        Load saved game

        Args:
            save_name: Name of save file

        Returns:
            Saved state or None
        """
        save_path = self.saves_dir / f"{save_name}.json"
        return self.load_json(save_path)

    def list_saves(self) -> list:
        """List available save files"""
        if not self.saves_dir.exists():
            return []
        return [f.stem for f in self.saves_dir.glob("*.json")]

    def delete_save(self, save_name: str) -> bool:
        """Delete a save file"""
        save_path = self.saves_dir / f"{save_name}.json"
        try:
            save_path.unlink()
            return True
        except Exception as e:
            self.logger.error(f"Error deleting save {save_name}: {e}")
            return False

File: sagacraft_rs/src/pyport/data/io_service.rs
use serde_json::Value;
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};

use crate::pyport::core::services::Service;

pub struct IOService {
    base_dir: PathBuf,
    adventures_dir: PathBuf,
    saves_dir: PathBuf,
}

impl IOService {
    pub fn new(base_dir: impl Into<PathBuf>) -> Self {
        let base_dir = base_dir.into();
        let adventures_dir = base_dir.join("adventures");
        let saves_dir = base_dir.join("saves");
        Self {
            base_dir,
            adventures_dir,
            saves_dir,
        }
    }

    pub fn load_json(&self, file_path: &Path) -> Result<Option<HashMap<String, Value>>, String> {
        match fs::read_to_string(file_path) {
            Ok(content) => match serde_json::from_str(&content) {
                Ok(data) => Ok(Some(data)),
                Err(e) => {
                    eprintln!("Invalid JSON in {:?}: {}", file_path, e);
                    Ok(None)
                }
            },
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
                eprintln!("File not found: {:?}", file_path);
                Ok(None)
            }
            Err(e) => Err(format!("Error loading {:?}: {}", file_path, e)),
        }
    }

    pub fn save_json(&self, file_path: &Path, data: &HashMap<String, Value>, indent: usize) -> Result<bool, String> {
        if let Some(parent) = file_path.parent() {
            fs::create_dir_all(parent).map_err(|e| format!("create dir {:?}: {}", parent, e))?;
        }
        let content = serde_json::to_string_pretty(data).map_err(|e| format!("serialize json: {}", e))?;
        fs::write(file_path, content).map_err(|e| format!("write file {:?}: {}", file_path, e))?;
        Ok(true)
    }

    pub fn load_adventure(&self, adventure_name: &str) -> Result<Option<HashMap<String, Value>>, String> {
        let adventure_path = self.adventures_dir.join(format!("{}.json", adventure_name));
        self.load_json(&adventure_path)
    }

    pub fn list_adventures(&self) -> Result<Vec<String>, String> {
        if !self.adventures_dir.exists() {
            return Ok(vec![]);
        }
        let mut adventures = vec![];
        for entry in fs::read_dir(&self.adventures_dir).map_err(|e| format!("read dir {:?}: {}", self.adventures_dir, e))? {
            let entry = entry.map_err(|e| format!("read entry: {}", e))?;
            let path = entry.path();
            if path.extension().and_then(|s| s.to_str()) == Some("json") {
                if let Some(stem) = path.file_stem().and_then(|s| s.to_str()) {
                    adventures.push(stem.to_string());
                }
            }
        }
        Ok(adventures)
    }

    pub fn save_game(&self, save_name: &str, game_state: &HashMap<String, Value>) -> Result<bool, String> {
        let save_path = self.saves_dir.join(format!("{}.json", save_name));
        self.save_json(&save_path, game_state, 2)
    }

    pub fn load_game(&self, save_name: &str) -> Result<Option<HashMap<String, Value>>, String> {
        let save_path = self.saves_dir.join(format!("{}.json", save_name));
        self.load_json(&save_path)
    }

    pub fn list_saves(&self) -> Result<Vec<String>, String> {
        if !self.saves_dir.exists() {
            return Ok(vec![]);
        }
        let mut saves = vec![];
        for entry in fs::read_dir(&self.saves_dir).map_err(|e| format!("read dir {:?}: {}", self.saves_dir, e))? {
            let entry = entry.map_err(|e| format!("read entry: {}", e))?;
            let path = entry.path();
            if path.extension().and_then(|s| s.to_str()) == Some("json") {
                if let Some(stem) = path.file_stem().and_then(|s| s.to_str()) {
                    saves.push(stem.to_string());
                }
            }
        }
        Ok(saves)
    }

    pub fn delete_save(&self, save_name: &str) -> Result<bool, String> {
        let save_path = self.saves_dir.join(format!("{}.json", save_name));
        match fs::remove_file(&save_path) {
            Ok(()) => Ok(true),
            Err(e) => {
                eprintln!("Error deleting save {}: {}", save_name, e);
                Ok(false)
            }
        }
    }
}

impl Service for IOService {
    fn name(&self) -> &'static str {
        "IOService"
    }

    fn initialize(&mut self, config: &HashMap<String, Value>) -> Result<(), String> {
        // Create directories
        fs::create_dir_all(&self.adventures_dir).map_err(|e| format!("create adventures dir: {}", e))?;
        fs::create_dir_all(&self.saves_dir).map_err(|e| format!("create saves dir: {}", e))?;

        // Override paths from config
        if let Some(Value::String(adventures_dir)) = config.get("adventures_dir") {
            self.adventures_dir = PathBuf::from(adventures_dir);
        }
        if let Some(Value::String(saves_dir)) = config.get("saves_dir") {
            self.saves_dir = PathBuf::from(saves_dir);
        }

        Ok(())
    }

    fn shutdown(&mut self) -> Result<(), String> {
        // Nothing needed for basic I/O
        Ok(())
    }
}